waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,605 @@
+                             {:instance-id id :killed? killed?
+                              :message message :service-id service-id :status status})]
+    ; request termination of the instance
+    (api-request http-client pod-url :request-method :delete :body term-json)
+    ; scale down the replicaset to reflect removal of this instance
+    (scale-service-by-delta scheduler service -1)
+    ; force-kill the instance (should still be terminating)
+    (api-request http-client pod-url :request-method :delete :body kill-json)
+    ; report back that the instance was killed successfully
+    (scheduler/process-instance-killed! instance)
+    200))
+
+(defn- service-spec
+  "Creates a Kubernetes ReplicaSet spec (with an embedded Pod spec) for the given Waiter Service."
+  [{:keys [rs-spec-file-path pod-base-port] :as scheduler} service-id service-description service-id->password-fn]
+  (let [{:strs [backend-proto cmd cpus grace-period-secs health-check-interval-secs
+                health-check-max-consecutive-failures mem min-instances ports run-as-user]} service-description
+        home-path (str "/home/" run-as-user)
+        common-env (scheduler/environment service-id service-description
+                                          service-id->password-fn home-path)
+        port0 pod-base-port
+        template-env (into [;; We set these two "MESOS_*" variables to improve interoperability
+                            {:name "MESOS_DIRECTORY" :value home-path}
+                            {:name "MESOS_SANDBOX" :value home-path}]
+                           (concat
+                             (for [[k v] common-env]
+                               {:name k :value v})
+                             (for [i (range ports)]
+                               {:name (str "PORT" i) :value (str (+ port0 i))})))
+        params {:k8s-name (service-id->k8s-name scheduler service-id)
+                :backend-protocol backend-proto
+                :backend-protocol-caps (string/upper-case backend-proto)
+                :cmd cmd
+                :cpus cpus
+                :env template-env
+                :grace-period-secs grace-period-secs
+                :health-check-interval-secs health-check-interval-secs
+                :health-check-max-consecutive-failures health-check-max-consecutive-failures
+                :health-check-url (sd/service-description->health-check-url service-description)
+                :home-path home-path
+                :memory  (str mem "Mi")
+                :min-instances min-instances
+                :port-count ports
+                :run-as-user run-as-user
+                :service-id service-id
+                :ssl? (= "https" backend-proto)}
+        edn-opts {:readers {'waiter/param params
+                            'waiter/param-str (comp str params)
+                            'waiter/port #(+ port0 %)
+                            'waiter.fn/into (fn [[xs ys]] (into xs ys))
```

Comment:
Are these `waiter.fn`s used? If not, we should remove them.

Reply:
Done.